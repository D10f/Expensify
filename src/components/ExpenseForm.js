import React from 'react'
import moment from 'moment'
import { SingleDatePicker } from 'react-dates'
import 'react-dates/initialize'
// import 'react-dates/lib/css/_datepicker.css' had to create a separate css file for this

class ExpenseForm extends React.Component {
  constructor(props){
    super(props)
    this.state = {
      description: props.expense ? props.expense.description : '',
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      notes: props.expense ? props.expense.notes : '',
      calendarFocused: false,
      error: ''
    }
  }

  onInputChange = (e) => {
    const name = e.target.name
    const value = e.target.value
    this.setState(() => ({ [name]: value }))
  }
  onAmountChange = (e) => {
    const amount = e.target.value

    if (!amount || amount.match(/^\d+(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }))
    }
  }
  onDateChange = (createdAt) => {
    if (createdAt){
      this.setState(() => ({ createdAt }))
    }
  }
  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }))
  }
  onFormSubmit = (e) => {
    e.preventDefault()

    if (!this.state.description || !this.state.amount){
      this.setState(() => ({ error: 'Please provide a description and amount for this expense.' }))
    } else {
      this.setState(() => ({ error: '' }))
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(), // to return time in ms
        notes: this.state.notes
      })
    }
  }
  render(){
    return (
      <form onSubmit={this.onFormSubmit}>
        {this.state.error && <p>{this.state.error}</p>}
        <label style={{ display: 'flex', flexDirection: 'column' }}>
        <b>Description</b>
          <input
            onChange={this.onInputChange}
            value={this.state.description}
            name="description"
            type="text"
            placeholder="e.g., Lunch with friends"
            autoFocus
          />
        </label>
        <label>
          <b>Amount</b>
          <input
            onChange={this.onAmountChange}
            value={this.state.amount}
            name="amount"
            type="text"
            placeholder="e.g.: 2.35"
          />
        </label>
        <SingleDatePicker
          date={this.state.createdAt}
          onDateChange={this.onDateChange}
          focused={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
          displayFormat="DD/MM/YYYY"
        />
        <textarea
          onChange={this.onInputChange}
          value={this.state.notes}
          name="notes"
          placeholder="Add a note for this expense (optional)"></textarea>
        <button>{this.props.expense ? 'Update Expense' : 'Create New Expense'}</button>
      </form>
    )
  }
}

export default ExpenseForm
