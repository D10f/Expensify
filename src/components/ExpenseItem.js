import React from 'react'
import { Link } from 'react-router-dom'
import numeral from 'numeral'

export const ExpenseItem = ({ id, description, amount, dispatch }) => (
  <li>
    <h3>{description}</h3>
    <Link to={`/edit/${id}`}>Edit</Link>
    <p>{numeral(amount / 100).format('$0,0.00')}</p>
  </li>
)

export default ExpenseItem
