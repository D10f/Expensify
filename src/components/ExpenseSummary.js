import React from 'react'
import { connect } from 'react-redux'
import numeral from 'numeral'
import selectExpenses from '../redux/selectors/expenses'
import selectTotalExpenses from '../redux/selectors/expenses-total'

export const ExpenseSummary = ({ expensesCount, expensesTotal }) => {
  const expenseWord = expensesCount === 1 ? 'expense' : 'expenses'
  const formattedExpensesTotal = numeral(expensesTotal / 100).format('$0,0.00')
  return (
    <header>
      <p>Viewing {expensesCount} {expenseWord} totalling {formattedExpensesTotal}</p>
    </header>
  )
}
const mapStateToProps = (state) => {
  const expensesCount = selectExpenses(state.expenses, state.filters)
  const expensesTotal = selectTotalExpenses(expensesCount)
  return {
    expensesCount: expensesCount.length,
    expensesTotal
  }
}

export default connect(mapStateToProps)(ExpenseSummary)
