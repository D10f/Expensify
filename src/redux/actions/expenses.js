import { v4 as uuid } from 'uuid'

export const addExpense = ({ id = uuid(), description = '', notes = '', amount = 100, createdAt = 0 } = {}) => ({
  type: 'ADD_EXPENSE',
  expense: {
    id,
    description,
    notes,
    amount,
    createdAt
  }
})

export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
})

export const editExpense = (id, updates = {}) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
})
