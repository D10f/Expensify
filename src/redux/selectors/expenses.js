import moment from 'moment'

export default (expenses, { text, startDate, endDate, sortBy }) => {
  return expenses.filter(expense => {
    const createdAtMoment = moment(expense.createdAt)
    const startDateMatch = startDate ? createdAtMoment.isSameOrAfter(startDate, 'day') : true
    const endDateMatch = endDate ? createdAtMoment.isSameOrBefore(endDate, 'day') : true
    // const startDateMatch = typeof startDate !== 'number' || expense.createdAt > startDate
    // const endDateMatch = typeof endDate !== 'number' || expense.createdAt < endDate
    const textMatch = expense.description.toLowerCase().includes(text.toLowerCase())

    return startDateMatch && endDateMatch && textMatch
  }).sort((a, b) => {
    if (sortBy === 'date') {
      return a.createdAt < b.createdAt ? 1 : -1
    }

    if (sortBy === 'amount') {
      return a.amount < b.amount ? 1 : -1
    }
  })
}
