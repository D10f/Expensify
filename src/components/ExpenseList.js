import React from 'react'
import { connect } from 'react-redux'
import selectExpenses from '../redux/selectors/expenses'
import ExpenseItem from './ExpenseItem'

export const ExpenseList = ({ expenses }) => (
  <ul>
    {
      expenses.length === 0
      ? (<li>No expenses found</li>)
      : ( expenses.map(expense => (
          <ExpenseItem key={expense.id} {...expense} />
        )))
    }
  </ul>
)

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  }
}

export default connect(mapStateToProps)(ExpenseList)
