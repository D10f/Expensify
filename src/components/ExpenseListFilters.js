import React from 'react'
import { DateRangePicker } from 'react-dates'
import 'react-dates/initialize'
import { connect } from 'react-redux'
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../redux/actions/filters'

export class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null
  }
  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value)
  }
  onSortChange = (e) => {
    e.target.value === 'date' ? this.props.sortByDate() : this.props.sortByAmount()
  }
  onFocusChange = (calendarFocused) => {
    this.setState(() => ({ calendarFocused }))
  }
  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate)
    this.props.setEndDate(endDate)
  }
  render(){
    return (
      <div>
        <input value={this.props.filters.text} onChange={this.onTextChange}/>
        <select value={this.props.filters.sortBy} onChange={this.onSortChange}>
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DateRangePicker
          startDate={this.props.filters.startDate}
          startDateId="your_unique_start_date_id"
          endDate={this.props.filters.endDate}
          endDateId="your_unique_end_date_id"
          focusedInput={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          onDatesChange={this.onDatesChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
          showClearDates={true}
          displayFormat="DD/MM/YYYY"
        />
      </div>
    )
  }
}

const mapStateToProps = (state) => ({
  filters: state.filters
})

const mapDispatchToProps = (dispatch) => ({
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate))
})

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters)
