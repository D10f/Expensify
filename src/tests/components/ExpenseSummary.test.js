import React from 'react'
import { shallow } from 'enzyme'
import { ExpenseSummary } from '../../components/ExpenseSummary'
import expenses from '../fixtures/expenses'

let wrapper

beforeEach(() => {
  wrapper = shallow(<ExpenseSummary />)
})

test('Should render ExpenseSummary correctly', () => {
  expect(wrapper).toMatchSnapshot()
})
