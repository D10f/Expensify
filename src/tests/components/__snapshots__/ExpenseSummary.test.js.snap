// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ExpenseSummary correctly 1`] = `
<header>
  <p>
    Viewing 
     
    expenses
     totalling 
    $0.00
  </p>
</header>
`;
